design-proposals/shadow-node.md

```diff
@@ -0,0 +1,180 @@
+  * Old virt-handler: during upgrade, the old RBACs are kept as backup until upgrade is done. Hence, patching node should not fail so there is no upgrade issue.
+  * Old virt-controllers: no upgrade issue, virt controller RBAC is not changed due to this solution.
+  * Old Virt-API: no upgrade issue.
+* New CRD; new RBAC; new virt-handler; old virt-controller; old virt-API:
+  * new virt-handler: shadowNode is written to but it is not synced to the node yet. However, virt-handler keeps patching the node directly, so there is no upgrade issue.
+  * [same as before] Old controllers: no upgrade issue, virt controller RBAC is not changed.
+  * [same as before] Old Virt-API: no upgrade issue.
+* New CRD; new RBAC; new virt-handler; new virt-controller; old virt-API:
+  * New virt-handler: no upgrade issue. virt-handler patches both node and shadow node.
+  * New controllers: will start getting shadowNode requests once they are created. no upgrade issue.
+  * [same as before] Old Virt-API: no upgrade issue.
+
+### After upgrade rollout:
+Virt-handler will keep trying (and failing) to patch the node since the backup RBACs are now removed. Virt-handler will ignore these cases by ignoring RBAC "forbidden" errors. This behavior should be kept until we no longer support upgrades from the current release.
+
+## Alternative solutions
+### Using status section
+Same proxy solution as primary solution where virt-hadler updates the labels/annotations somewhere, and it gets copied to the node by virt-controller, but instead of using a new CRD as the medium - writing the labels/Annotations to the status section of an existing CRD like kubevirt CR.
+
+Pros:
+- Avoid introducing a new CRD to the project.
+
+Cons:
+- The status section is considered a user facing API section. Dynamically adding/updating 150+ technical labels there may only confuse the user and thus is not advisable.
+
+### Using the virt-handler pod as a medium to pass the labels/annotations
+Same proxy solution as primary solution where virt-handler updates the labels/annotations somewhere, and it gets copied to the node by virt-controller, but instead of using a new CRD as the medium - writing the labels/annotations to the virt-handler pods's labels/annotations.
+Labels/Annotations that need to pass to the node should get a special prefix (like "nodesync/"), so that regular labels don't get passed to the node by mistake.
+
+Pros:
+- Avoid introducing a new CRD to the project.
+- A simple and clean approach.
+
+Cons:
+- It requires giving the virt-handler RBAC permission to update pods, which is generally considered risky.
+- It's only relevant for passing labels/annotations. CRD approach allows for future expansions.
+
+### Creating a GRPC communication between virt-handler and virt-controller
+Pass the labels/annotations through a grpc channel to the virt-controller.
+
+Pros:
+- Avoid introducing a new CRD to the project.
+- Allows for future expansions.
```

Comment:
Do you have specific expansions in mind? This statement seems generic.

I would suggest to add here a security benefit: virt-handler does not need to access kube api at all, and cannot modify unrelated nodes or shadow nodes. virt-controller does all the risky work.

Reply:
>Do you have specific expansions in mind? This statement seems generic.

it is. Will remove.

>I would suggest to add here a security benefit: virt-handler does not need to access kube api at all, and cannot modify unrelated nodes or shadow nodes. virt-controller does all the risky work.

Added.